Packages/nadena.dev.modular-avatar/Editor/OptimizationPasses/GCGameObjectsPass.cs

```diff
@@ -23,11 +31,132 @@ internal GCGameObjectsPass(BuildContext context, GameObject root)
 
         internal void OnPreprocessAvatar()
         {
+            CollectReferences();
+            Cleanup();
             MarkAll();
             Sweep();
+            Debug.Log($"Cleanup complete, removed {_removedObjects} objects and {_removedComponents} components.");
         }
 
+        private void CollectReferences()
+        {
+            LoopAllComponents((obj, component) =>
+            {
+                switch (component)
+                {
+                    case SkinnedMeshRenderer mesh:
+                        CollectBonesFromSkinnedMeshRenderer(mesh);
+                        break;
+
+                    case VRCPhysBone pb:
+                        _physBones.Add(pb);
+                        break;
+                    case VRCPhysBoneCollider collider:
+                        _physBoneColliders.Add(collider);
+                        break;
+                }
+            });
+        }
+
+        private void CollectBonesFromSkinnedMeshRenderer(SkinnedMeshRenderer renderer)
+        {
+            HashSet<int> usedBonesIndices = new HashSet<int>();
+            foreach (var boneWeight in renderer.sharedMesh.boneWeights)
+            {
+                // technically should add only if weight is > 0, but not worth the cost and does not have negative effects
+                usedBonesIndices.Add(boneWeight.boneIndex0);
+                usedBonesIndices.Add(boneWeight.boneIndex1);
+                usedBonesIndices.Add(boneWeight.boneIndex2);
+                usedBonesIndices.Add(boneWeight.boneIndex3);
+            }
+
+            Transform[] bones = renderer.bones;
+            foreach (var usedBoneIndex in usedBonesIndices)
+            {
+                Transform usedBone = bones[usedBoneIndex];
+                if (usedBone != null) _usedBones.Add(usedBone.gameObject);
+            }
+
+            foreach (var bone in bones)
+            {
+                if (bone != null) _allBones.Add(bone.gameObject);
+            }
+        }
+
+        private void Cleanup()
+        {
+            // remove phys bones that only affect bones that are not used
+            var bonesToRemove = new HashSet<VRCPhysBone>();
+            foreach (var vrcPhysBone in _physBones)
+            {
+                vrcPhysBone.InitTransforms();
+                if (vrcPhysBone.bones.Any(x => _usedBones.Contains(x.transform.gameObject))) continue;
+                // in case someone uses the component on chain of something that is not actually a bone
+                if (vrcPhysBone.bones.Any(x => !_allBones.Contains(x.transform.gameObject))) continue;
+                bonesToRemove.Add(vrcPhysBone);
+            }
```

Comment:
PhysBones can influence other things than just mesh renderers - lights, particle systems, audio sources, colliders, etc. We need to take this into account before deleting physbones here.

Reply:
Thats why there is second check for "all bones", if it contains anything that is not a bone it wlll be left alone.

Reply:
To clarify - if you have a bone which has, say, a light on it, and which is also in a renderer's bones array (but has no vertices weighted to it), then it will be in _allBones, but not in _usedBones. This in turn means a physbone that influences a bone like this can be dropped.

Reply:
in this case yes, but it must be actual bone, added in blender, and thats rather weird use case - not even use how it would work with a blender bone not weighted to anything, would probabvly require separate way to flag such weird cases to not be removed - as overall I want to remove objects like that as hard to tell apart bone that is no longer used because object is gone vs someone doing it on purpose. 
As if the skirt contained some additional components on it like lights or particle systems, they should be removed too.